data/Builder.go

```diff
@@ -0,0 +1,110 @@
+package data
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"strings"
+)
+
+//Datum represent one data point
+type Datum map[string]interface{}
+
+//Dataset represents an array of data points
+type Dataset []Datum
+
+//GetSelector returns the selector fields for a given platform datatype, or nil if there is no match
+func GetSelector(data interface{}) interface{} {
+
+	switch data.(type) {
+	case *Basal:
+		return data.(*Basal).Selector()
+	case *DeviceEvent:
+		return data.(*DeviceEvent).Selector()
+	default:
+		return nil
+	}
+}
+
+//Builder interface that the TypeBuilder implements
+type Builder interface {
+	BuildFromRaw(raw []byte) (interface{}, *Error)
+	BuildFromData(data map[string]interface{}) (interface{}, *Error)
+	BuildFromDataSet(dataSet Dataset) ([]interface{}, *ErrorSet)
+}
+
+//TypeBuilder that is used to build data types that the platform understands
+type TypeBuilder struct {
+	inject map[string]interface{}
+}
+
+//NewTypeBuilder returns an instance of TypeBuilder
+func NewTypeBuilder(inject map[string]interface{}) Builder {
+	return &TypeBuilder{
+		inject: inject,
+	}
+}
+
+//BuildFromDataSet will build the matching type(s) from the given Dataset
+func (typeBuilder *TypeBuilder) BuildFromDataSet(dataSet Dataset) ([]interface{}, *ErrorSet) {
+
+	var set []interface{}
+	var buildError *ErrorSet
+
+	for i := range dataSet {
+		item, err := typeBuilder.BuildFromData(dataSet[i])
+		if err != nil && !err.IsEmpty() {
+			if buildError == nil {
```

Comment:
Rather than check this every time through the loop, perhaps create at beginning?  Then, if we are going to stick with Error.IsEmpty perhaps add an ErrorSet.IsEmpty and use that at the end of this function on whether to return the ErrorSet or nil (if the ErrorSet is empty).